### DISCUSSION TITLE
Codegen question

### DISCUSSION THREAD

**Discussion by buhaytza2005** [sreeise/graph-rs-sdk#463] [General]
When doing a regeneration of `me` - trying some stuff out with the bookings - I came across a couple of bits:
 
- DefaultDriveApi is no longer imported - am I correct in assuming that was added manually at some point? (I can add to the settings)
- Does the renaming of the methods happen automatically on rewrite or should the macro be invoked after the generation? (some of the methods don't get renamed and I can't really tell how to sort it out)
- Is there a way to add a module to Graph through codegen or is that a manual entry?

Secondly, bookings have the following paths: 

```
"/solutions"
"/solutions/bookingBusinesses"
"/solutions/bookingBusinesses/$count"
"/solutions/bookingBusinesses/{bookingBusiness-id}"
"/solutions/bookingBusinesses/{bookingBusiness-id}/appointments"
[...]
```
I can see the `Solutions` have been added as part of the default resource identities.

Would this then be dropped in `src` and hooked up to Graph? Or would it make more sense to have it linked to `me` and `users`?

**Comment by sreeise** [comment]
> When doing a regeneration of `me` - trying some stuff out with the bookings - I came across a couple of bits:
> 
>     * DefaultDriveApi is no longer imported - am I correct in assuming that was added manually at some point? (I can add to the settings)


Hey @buhaytza2005 

The Default drive is one that is done manually. I may be mistaken, but I believe this is due to how the methods for drive that have a /user or a /me path are different and they may not have been in the OpenAPI specifically. 

If your talking about regenerating the main me client does not import that specific module then yeah there is place to add that. 

Looking at it you are right it is not bring imported when it should be. 

**https://github.com/sreeise/graph-rs-sdk/blob/b853c90c7d535856fe61413fa5987f005c558e3c/graph-codegen/src/settings/resource_settings.rs#L1183-L1184**

The `ResourceSettings` is what controls the imports but I have been trying to move away from that for a while. . If you just add a star import for default drives: `use crate::default_drive::*` to the place above then it will be generated.  

One should be added for `users` as well https://github.com/sreeise/graph-rs-sdk/blob/b853c90c7d535856fe61413fa5987f005c558e3c/graph-codegen/src/settings/resource_settings.rs#L1177-L1178

> Does the renaming of the methods happen automatically on rewrite or should the macro be invoked after the generation? (some of the 
> methods don't get renamed and I can't really tell how to sort it **out)**

If your referring to how you would need to add configuration for renaming methods then that would happen by rerunning code generation.
Does that answer your question? If not let me know.


> Is there a way to add a module to Graph through codegen or is that a manual entry?

The code generation handles creating the underlying directory and module for a resource but you still have to add the name of the module to src/lib.rs for it to compile.

> Secondly, bookings have the following paths:
> 
> ```
> "/solutions"
> "/solutions/bookingBusinesses"
> "/solutions/bookingBusinesses/$count"
> "/solutions/bookingBusinesses/{bookingBusiness-id}"
> "/solutions/bookingBusinesses/{bookingBusiness-id}/appointments"
> [...]
> ```
> 
> I can see the `Solutions` have been added as part of the default resource identities.
> 
> Would this then be dropped in `src` and hooked up to Graph? Or would it make more sense to have it linked to `me` and `users`?

Because the solutions paths begin with solutions then it would be its own module. It would only be under me or users if the path started with /me or /users.

You would run the codegen, add a `pub mod solutions;` to src/lib.rs, and then in src/client/graph.rs you would add a macro call in the main graph impl to create the method for solutions. From what I can tell `solutions` does not have an id directory after `/solutions` in its path so codgen would generate a single struct and then in src/client/graph.rs you would add the macro call similar to:

```rust
api_client_impl!(solutions, SolutionsApiClient);
```

**Reply by buhaytza2005** [reply]
I started to work on this and got most of the `bookingBusinesses` working.

I have come across something that gives me pause.
```
"/solutions/bookingBusinesses/{bookingBusiness-id}/calendarView"
"/solutions/bookingBusinesses/{bookingBusiness-id}/calendarView/$count"
"/solutions/bookingBusinesses/{bookingBusiness-id}/calendarView/{bookingAppointment-id}"
"/solutions/bookingBusinesses/{bookingBusiness-id}/calendarView/{bookingAppointment-id}/cancel"
```

There already is a `CalendarView` value in the `ResourceIdentity` enum. Can I just reuse and link to `bookingBusiness` or would I need to do something else to use a different name?

**Reply by sreeise** [reply]
Interesting. I think in this case these are methods specific to the bookingBusinesses CalendarView right? Meaning if you were to just look up all the APIs with CalendarView, are these methods also showing up for users or say groups (resources who have calendarView APIs)? 

 If I run this in an example file in the repository:
 
 ```rust
use graph_codegen::openapi::OpenApi;

fn main() {
    let open_api = OpenApi::default();

    open_api.debug_path_contains("calendarView");
}
```

I don't see any other resource with those two specific paths. So in this case, if these 4 paths are the only ones related to calendarView I think you could just allow these methods to be generated on a BookingBusinesses client. Does that make sense?

As far as using the ResourceIdentity::CalendarView, I think that would probably interfer with code generation elsewhere but you could try it potentially. Basically the codegen will use the ResourceIdentity to add methods to the impl of a resource such as links between different resources. And in this case if you use the calendar view variant already there then it would add a lot of unwanted methods. 

If you keep it under the BookingBusinesses resource you woudn't need to add any variant at all so that might work better. Or under SolutionsApiClient if there are relatively few specific booking businesses paths.
